src/phoenix/utilities/ratelimits.py

```diff
@@ -0,0 +1,210 @@
+P = ParamSpec("P")
+T = TypeVar("T")
+A = TypeVar("A", bound=Callable[..., Awaitable[Any]])
+
+
+class UnavailableTokensError(Exception):
+    pass
+
+
+class TokenLimiter:
+    """
+    A simple in-memory rate-limiter implemented using the leaky bucket algorithm.
+    """
+
+    def __init__(
+        self,
+        per_minute_rate: Numeric,
+        starting_tokens: Numeric,
+        max_tokens: Numeric,
+        rate_multiplier: Numeric = 1,
+    ):
+        self.rate_multiplier = rate_multiplier
+        self.rate = self._per_second_rate(per_minute_rate)
+        self.tokens = starting_tokens
+        self.max_tokens = max_tokens
+        self.created = time.time()
+        self.last_checked = self.created
+        self.total_tokens: Numeric = 0
+
+    def _per_second_rate(self, per_minute_rate: Numeric) -> float:
+        return round(per_minute_rate / 60, 3) * self.rate_multiplier
+
+    def refresh_limit(self, per_minute_rate: Numeric, max_tokens: Numeric) -> None:
+        new_rate = self._per_second_rate(per_minute_rate)
+        if self.rate != new_rate:
+            self.rate = new_rate
+            self.tokens = 0  # reset tokens as conservatively as possible
+            self.max_tokens = max_tokens
+            self.last_checked = time.time()
+
+    def available_tokens(self) -> float:
+        time_since_last_checked = time.time() - self.last_checked
+        return min(self.max_tokens, self.rate * time_since_last_checked + self.tokens)
+
+    def spend_tokens_if_available(self, token_cost: Numeric) -> None:
+        if token_cost > self.available_tokens():
+            raise UnavailableTokensError
+        now = time.time()
+        current_tokens = min(self.max_tokens, self.tokens + (now - self.last_checked) * self.rate)
+        self.tokens = current_tokens - token_cost
+        self.last_checked = now
+        self.total_tokens += token_cost
+
+    def spend_tokens(self, token_cost: Numeric) -> None:
+        self.tokens -= token_cost
+        self.total_tokens += token_cost
+
+    def wait_for_then_spend_available_tokens(self, token_cost: Numeric) -> None:
+        MAX_WAIT_TIME = 5 * 60  # 5 minutes
+        start = time.time()
+        while (time.time() - start) < MAX_WAIT_TIME:
+            try:
+                self.spend_tokens_if_available(token_cost)
+                break
+            except UnavailableTokensError:
+                time.sleep(1 / self.rate)
+                continue
+
+    async def async_wait_for_then_spend_available_tokens(self, token_cost: Numeric) -> None:
+        MAX_WAIT_TIME = 5 * 60
+        start = time.time()
+        while (time.time() - start) < MAX_WAIT_TIME:
+            try:
+                self.spend_tokens_if_available(token_cost)
+                break
+            except UnavailableTokensError:
+                await asyncio.sleep(1 / self.rate)
+                continue
+
+    def effective_rate(self) -> Numeric:
+        return self.total_tokens / (time.time() - self.created)
```

Comment:
this is just going to make the effective rate continue to get smaller. best to keep this side-effect free.

Reply:
Is this even needed?

Reply:
`total_tokens` increases over time as the rate limiter is used, it's only useful if anyone ever wants to inspect how what the observed token rate over the lifetime of the rate limiter is. We can axe this